x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,836 @@
+
+	stopTimeToConsumerIdsKeyPrefix := types.StopTimeToConsumerIdsKeyPrefix()
+	iterator := storetypes.KVStorePrefixIterator(store, []byte{stopTimeToConsumerIdsKeyPrefix})
+	defer iterator.Close()
+
+	result := []string{}
+	for ; iterator.Valid(); iterator.Next() {
+		stopTime, err := types.ParseTime(types.StopTimeToConsumerIdsKeyPrefix(), iterator.Key())
+		if err != nil {
+			k.Logger(ctx).Error("failed to parse stop time",
+				"error", err)
+			continue
+		}
+		if stopTime.After(ctx.BlockTime()) {
+			return result
+		}
+
+		consumerIds, err := k.GetConsumersToBeStopped(ctx, stopTime)
+		if err != nil {
+			k.Logger(ctx).Error("failed to retrieve consumers to stop",
+				"stop time", stopTime,
+				"error", err)
+			continue
+		}
+		if len(result)+len(consumerIds) >= int(limit) {
+			remainingConsumerIds := len(result) + len(consumerIds) - int(limit)
+			if len(consumerIds[:len(consumerIds)-remainingConsumerIds]) == 0 {
+				return result
+			}
+			return append(result, consumerIds[:len(consumerIds)-remainingConsumerIds]...)
+		} else {
+			result = append(result, consumerIds...)
+		}
+
+		consumerIds = append(consumerIds, string(iterator.Key()[1+8:]))
+	}
+
+	return result
+}
+
+// IsValidatorOptedInToChainId checks if the validator with `providerAddr` is opted into the chain with the specified `chainId`.
+// It returns `found == true` and the corresponding chain's `consumerId` if the validator is opted in. Otherwise, it returns an empty string
+// for `consumerId` and `found == false`.
+func (k Keeper) IsValidatorOptedInToChainId(ctx sdk.Context, providerAddr types.ProviderConsAddress, chainId string) (string, bool) {
+	consumerIds, err := k.GetOptedInConsumerIds(ctx, providerAddr)
+	if err != nil {
+		k.Logger(ctx).Error("failed to retrieve the consumer ids this validator is opted in to",
+			"providerAddr", providerAddr,
+			"error", err)
+		return "", false
+	}
+
+	for _, consumerId := range consumerIds {
+		consumerChainId, err := k.GetConsumerChainId(ctx, consumerId)
+		if err != nil {
+			k.Logger(ctx).Error("cannot find chain id",
+				"consumerId", consumerId,
+				"error", err)
+			continue
+		}
+
+		if consumerChainId == chainId {
+			return consumerId, true
+		}
+
+	}
+	return "", false
+}
+
+func (k Keeper) PrepareConsumerForLaunch(ctx sdk.Context, consumerId string, previousSpawnTime *time.Time, spawnTime time.Time) {
+	if previousSpawnTime != nil {
+		// if this is not the first initialization, remove the consumer id from the old spawn time
+		k.RemoveConsumerFromToBeLaunchedConsumers(ctx, consumerId, *previousSpawnTime)
+	}
+
+	k.AppendSpawnTimeForConsumerToBeLaunched(ctx, consumerId, spawnTime)
+}
+
+// CanLaunch checks on whether the consumer with `consumerId` has set all the initialization parameters set and hence
+// is ready to launch
+func (k Keeper) CanLaunch(ctx sdk.Context, consumerId string) bool {
+	if initializationParameters, err := k.GetConsumerInitializationParameters(ctx, consumerId); err != nil {
```

Comment:
shouldn't that be err==nil ?
Would be good to log an error in that case and for the positive case a info log would be good to see due to which 
parameter the chain is not ready to launch